### PR TITLE
[DTensor] Add `async_output` kwags to DTensor.to_local() API

### PR DESCRIPTION
**Pull request by yoyoyocmu**
Summary:
This change adds `async_output` kwags to DTensor.to_local() API and allow users to specify if they want the tensors to be synced in forward/backward path during `torch.Tensor` and `DTensor` conversion.

The `async_output` optional is already available in `_ToTorchTensor` and this change expose it to public API.

Before the change, we ran into two issues during training when converting part of the model to `DTensor`:
1. Numerical value is not correct in backward `_FromTorchTensor` collective functions.
2. Checkpointing issue in prodcution models.

Test Plan:
**CI**
CI test

**Unit test**:
`buck2 test mode/dev-nosan //caffe2/test/distributed/_tensor:dtensor`
- Passed

**WHEN model test**:
Tested together with: https://github.com/pytorch/pytorch/pull/115847
- Verified loss for each iteration within the expected range.
- Verified NE on-par with this change with 4B training data.

Differential Revision: D52173387

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/_tensor/api.py [hunk 1]
```diff
@@ -362,10 +374,14 @@ def from_local(
...
             stride,
+            async_output,
         )
 
     def to_local(
-        self, *, grad_placements: Optional[Sequence[Placement]] = None
+        self,
+        *,
+        grad_placements: Optional[Sequence[Placement]] = None,
+        async_output: str = True,
```
Comment (wanchaol) [inline torch/distributed/_tensor/api.py:384]: What's the concrete use case of having this exposed as a API, instead of doing something like below explicitly:

````
out = dtensor.to_local()
if isinstance(out, AsyncCollectiveTensor):
    out = out.wait()
```
We should be restrictive to what public API we expose so unless there's a strong reason to do this I would propose we do this manually

Comment (yoyoyocmu) [reply torch/distributed/_tensor/api.py:384]: Thanks for the suggestion. There're are two major places in the DTensor API where `to_local()` is called, one is the forward `DTensor.to_local()` and the other is the `DTensor.from_local()` backward. For the second one, any recommend way to do the check and wait() (i.e. in the backward hook)?

**Comment by yoyoyocmu** [comment]
Closing this PR as discussed offline, we could explicitly `wait()` from caller side via backward hook.
